File: archive/word-pattern.js
const wordPattern = function(pattern, s) {
  pattern = pattern.split('')
  s = s.split(' ')
  if (pattern.length !== s.length) {
    return false
  }

  let mem = {}
  let set = new Set()
  for (let i = 0; i < s.length; i++) {
    if (mem[pattern[i]]!==s[i]) {
      if (mem[pattern[i]] === undefined) {
        if (set.has(s[i])) {
          return false
        } else {
          mem[pattern[i]] = s[i]
          set.add(s[i])
        }
      } else {
        return false
      }
    }
  }

  return true
}

console.log(wordPattern("abba", "dog cat cat dog"))

File: go-archive/word-pattern.go
func wordPattern(pattern string, s string) bool {
    spl:= strings.Split(s, " ")
    p:=map[byte]string{}; r:=map[string]byte{}
    if (len(spl) != len(pattern)) {return false}
    for i:=0; i < len(pattern); i++ {
        if _, isHas:= p[pattern[i]]; isHas {
            if (p[pattern[i]] != spl[i]) {
                return false
            }
        } else {
            p[pattern[i]] = spl[i]
        }

        if _, isHas:= r[spl[i]]; isHas {
            if (r[spl[i]] != pattern[i]) {
                return false
            }
        } else {
            r[spl[i]] = pattern[i]
        }
    }

    return true
}